### ISSUE TITLE
There is no reasonable way to see all invalidated configuration cache inputs

### ISSUE THREAD

**Issue by mgroth0** [gradle/gradle#26440]
### Current Behavior

**TLDR: Current configuration cache reports and logs do not contain a list if  _all_ of the configuration cache inputs that invalidated the cache. They only show one, which makes it near impossible to debug situations where we need to see a large set of inputs that are invalidating the cache.**

I use a custom gradle plugin which is added as a classpath dependency in the buildscript block of settings.gradle.kts. This is a folder of jars which I update periodically.

Whenever I update these jars, the configuration cache is invalidated. This is expected.

However, the configuration cache is actually invalidated _twice_. This is not expected. 

As in, if I update my gradle settings plugin jars I then have to run two full builds in order to establish a working configuration cache. I should only have to do one.

According to the gradle info logs, the configuration cache inputs that are invalidating the build on the second time after updating my plugin jars are some internal system properties related to kotlin and/or intellij. For example:

- `project.structure.add.tools.jar.to.new.jdk`
- `psi.track.invalidation`
- `psi.incremental.reparse.depth.limit`
- ... and likely many more.

I have reported this issue on the JetBrains youtrack first. That report is [here](https://youtrack.jetbrains.com/issue/KT-62074/project.structure.add.tools.jar.to.new.jdk-is-wrongly-invalidating-the-gradle-configuration-cache). Originally, I viewed this as something that JetBrains was primarily responsible for fixing.

But then I though I could implement a workaround. I thought that maybe, if I could discover all of the properties that are causing this issue, I could set them all manually to prevent them from invalidating the configuration cache.

Here is where there is a problem on Gradle's end: there is no reasonable way to see all of the properties that are the issue. When the configuration cache invalidates, it only reports **one**  configuration cache input that caused the cache to invalidate. Even if there are more than one. Even on debug logging, it only reports a single input that caused the cache to invalidate. Given that there could be dozens of properties that I need to set manually, it becomes unrealistic for me to run dozens of builds to discover what configurration cache inputs are causng the issue.

### Expected Behavior

Configuration cache reports and/or logs should show all of inputs that caused invalidation, not just one.

### Context (optional)

In general, this makes the process of debugging the configuration cache much more difficult in any scenario when there are more than one problematic configuration cache input. If debugging also requires running long builds, it becomes close to impossible to debug if you have to run a new build each time to debug a single configuration cache input.

### Steps to Reproduce

1. Invalidate the configuration cache with multiple properties
2. Run a simple build, with debug logging.
3. Observe that in the entire debug log and configuration cache report, you can only find information on a single one of the configuration inputs that caused the invalidation.

### Gradle version

8.3

**Comment by jbartok** [comment]
Thank you for your interest in Gradle!

This feature request is in the backlog of the relevant team and is prioritized by them.

**Comment by mlopatkin** [comment]
Current implementation stops at the first detected discrepancy for performance - when you're not debugging the cache, you mostly care only if it matches or not. However, it definitely can be useful in certain cases. We may introduce this behavior under a flag in the future (no promises or timelines :)).

**Comment by adammurdoch** [comment]
We should try to avoid adding more flags unless they are really necessary.

Gradle could report all system properties whose values have invalidated the cache entry without any performance penalty. It has to check them all on a hit and the cost of checking them all on a miss is basically zero relative to the cost of configuring everything. Same applies to Gradle properties.

For files, Gradle can report changes to files on a miss as it snapshots them during configuration, again without any performance penalty.
